src/Domain/Item/ItemService.ts

```diff
@@ -49,8 +55,138 @@ export class ItemService implements ItemServiceInterface {
+      const existingItem = await this.itemRepository.findByUuidAndUserUuid(itemHash.uuid, dto.userUuid)
+
+      if (!this.itemShouldBeSaved(itemHash, existingItem)) {
+        conflicts.push({
+          serverItem: existingItem,
+          type: 'sync_conflict'
+        })
+
+        continue
+      }
+
+      if(existingItem) {
+        const updatedItem = await this.updateExistingItem(existingItem, itemHash, dto.userAgent)
+        savedItems.push(updatedItem)
+      } else {
+        try {
+          const newItem = await this.saveNewItem(itemHash, dto.userAgent)
+          savedItems.push(newItem)
+        } catch (error) {
+          conflicts.push({
+            unsavedItem: itemHash,
+            type: 'uuid_conflict',
+          })
+
+          continue
+        }
+      }
+    }
+
+    const syncToken = this.calculateSyncToken(lastUpdatedTimestamp, savedItems)
+
+    return {
+      savedItems,
+      conflicts,
+      syncToken
+    }
+  }
+
+  async frontLoadKeysItemsToTop(userUuid: string, retrievedItems: Array<Item>): Promise<Array<Item>> {
+    const itemsKeys = await this.itemRepository.findAll({
+      userUuid,
+      contentType: Item.CONTENT_TYPE_ITEMS_KEYS,
+      sortBy: 'updated_at_timestamp',
+      sortOrder: 'DESC'
+    })
+
+    const retrievedItemsIds: Array<string> = retrievedItems.map((item: Item) => item.uuid)
+
+    itemsKeys.forEach((itemKey: Item) => {
+      if(retrievedItemsIds.indexOf(itemKey.uuid) === -1) {
+        retrievedItems.unshift(itemKey)
+      }
+    })
+
+    return retrievedItems
+  }
+
+  private calculateSyncToken(lastUpdatedTimestamp: number, savedItems: Array<Item>): string {
+    if (savedItems.length) {
+      const sortedItems = savedItems.sort((itemA: Item, itemB: Item) => itemA.updatedAt > itemB.updatedAt ? 1 : -1)
+      lastUpdatedTimestamp = sortedItems[sortedItems.length - 1].updatedAt
+    }
+
+    const lastUpdatedTimestampWithMicrosecondPreventingSyncDoubles = lastUpdatedTimestamp + 1
+
+    return Buffer.from(
+      `${this.SYNC_TOKEN_VERSION}:${lastUpdatedTimestampWithMicrosecondPreventingSyncDoubles / 1000000}`,
+      'utf-8'
+    ).toString('base64')
+  }
+
+  private async updateExistingItem(existingItem: Item, itemHash: ItemHash, userAgent: string): Promise<Item> {
+    existingItem.content = itemHash.content
+    existingItem.contentType = itemHash.content_type
+    if (itemHash.deleted !== undefined) {
+      existingItem.deleted = itemHash.deleted
+    }
+    if (itemHash.duplicate_of) {
+      existingItem.duplicateOf = itemHash.duplicate_of
+    }
+    existingItem.encItemKey = itemHash.enc_item_key
+    existingItem.itemsKeyId = itemHash.items_key_id
+    existingItem.lastUserAgent = userAgent
+
+    if (itemHash.deleted === true) {
+      existingItem.deleted = true
+      existingItem.content = null
+      existingItem.encItemKey = null
+      existingItem.authHash = null
+    }
+
+    existingItem.createdAt = dayjs.utc(itemHash.created_at).valueOf() * 1000
```

Comment:
Why not use the same code block from `saveNewItem` here as well?

```
    const now = this.timer.getTimestampInMicroseconds()
    existingItem.createdAt = now
    existingItem.updatedAt = now
```

Reply:
Ohh wait nvm, what am I saying. This is an existing item. So then why are we overwriting createdAt with its same value? Is it essentially just to ensure it has the correct precision? But if we already did the migration when would it be the case that it has the wrong precision?

Reply:
Will remove overwritting `created_at`